Using extension mechanism to get the implementation for the HBaseDDLExecutor.

sagarkapare opened the pull request.

sagarkapare committed:
CDAP-7685 Using extension mechanism to get the implementation for the HBaseDDLExecutor.

sagarkapare commented:
Build is running here - https://builds.cask.co/browse/CDAP-DUT5316-1

albertshau commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java:
```diff
@@ -45,9 +45,9 @@
+  private Configuration hConf;
 
-  public DefaultHBaseDDLExecutor(Configuration hConf) {
+  void setHConf(Configuration hConf) {
```
does this mean extensions written by users won't have access to the Configuration? How will users be able to configure their classes? 

Perhaps we should have an initialize(Configuration hConf) method in the interface?

sagarkapare replied:
Adding `initialize(Configuration hConf)` will add dependency on the hadoop though. Should we have `Configuration` class in the spi as well? or should this method just take `Map<String, String> properties`, where properties will include properties from both `Configuration` and `CConfiguration`?

albertshau replied:
hm that's a good question. Let's do this in a separate PR

sagarkapare replied:
https://issues.cask.co/browse/CDAP-8190 to track.

albertshau replied:
one possibility is to pass in some HBaseDDLContext interface instead of Configuration, and the context has a `String getProperty(String key)` method, and a `<T> T getHadoopConfiguration()` method similar to what is done in MapReduceContext.getHadoopJob(). The implementation would just call get() on a CConfiguration for getProperty() and would just return the hConf for getHadoopConfiguration().

albertshau commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java:
```diff
@@ -239,8 +245,8 @@ public void createTableIfNotExists(HBaseAdmin admin, TableId tableId, HTableDesc
-    DefaultHBaseDDLExecutor executor = (DefaultHBaseDDLExecutor) getHBaseDDLExecutor(admin.getConfiguration());
-    TableDescriptor descriptor = executor.getTableDescriptor(tableDescriptor);
+    HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, admin.getConfiguration()).get();
+    TableDescriptor descriptor = ((DefaultHBaseDDLExecutor) executor).getTableDescriptor(tableDescriptor);
```
this isn't going to work with a custom implementation, though it seems like this will be addressed in a future PR.

sagarkapare replied:
yes this will be done in separate PR. where createTable method will directly accept `TableDescriptor` then this conversion is not needed.

albertshau commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java:
```diff
@@ -113,10 +114,15 @@
   protected NamespaceQueryAdmin namespaceQueryAdmin;
 
+  @Inject
+  protected HBaseDDLExecutorFactory hBaseDDLExecutorFactory;
```
is this intended? seems like it's not used.

sagarkapare replied:
yes this is not needed. will remove

albertshau reviewed (commented):
mostly commenting on adding an initialize method to the DDLExecutor interface.

sagarkapare commented:
@albertshau Addressed comments. Please take a look again. Also added second commit where for createTable method instead of passing `HBaseAdmin` passing `HBaseDDLExecutor`.

sagarkapare commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamConsumerStateStoreFactory.java:
```diff
@@ -54,7 +59,9 @@ public synchronized StreamConsumerStateStore create(StreamConfig streamConfig) t
+    try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get();
+         // TODO should tableExists be part of HBaseDDLExecutor??
+         HBaseAdmin admin = new HBaseAdmin(hConf)) {
       if (!tableUtil.tableExists(admin, hbaseTableId)) {
```
Since `tableExist` method is not part of the `HBaseDDLExecutor`, however we still want to reuse the `admin`, should `HBaseDDLExecutor` have `getAdmin` method?

albertshau replied:
shouldn't have a getAdmin, as that breaks the abstraction. You can probably just always call createTableIfNotExists, unless it's expensive to build the table descriptor. Otherwise, it's probably ok to have the 2 calls, though I would only create the DDLExecutor if the table doesn't exist.

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/kv/HBaseKVTableDefinition.java:
```diff
@@ -121,8 +125,8 @@ public void create() throws IOException {
 
-      try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
-        tableUtil.createTableIfNotExists(admin, tableId, tableDescriptor.build());
+      try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get()) {
```
seems like you can create the factory in the constructor.

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java:
```diff
@@ -552,8 +554,8 @@ private void createQueueTable(TableId tableId, HTableDescriptorBuilder htd, byte
 
-      try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
-        tableUtil.createTableIfNotExists(admin, tableId, htd.build(), splitKeys);
+      try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get()) {
```
can the factory be created once?

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/hbase/HBaseStreamFileConsumerFactory.java:
```diff
@@ -88,8 +90,8 @@ protected StreamConsumer create(TableId tableId, StreamConfig streamConfig, Cons
 
-    try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
-      tableUtil.createTableIfNotExists(admin, hBaseTableId, htd.build(), splitKeys,
+    try (HBaseDDLExecutor executor = new HBaseDDLExecutorFactory(cConf, hConf).get()) {
```
can the factory be created once in the constructor?

albertshau commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/DefaultHBaseDDLExecutor.java:
```diff
@@ -45,14 +45,27 @@
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> T getConf() {
```
why is this needed?

albertshau commented on cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java:
```diff
@@ -17,13 +17,29 @@
+   * @param <T>
+   * @return the conf
+   */
+  <T> T getConf();
```
why is this needed?

sagarkapare replied:
This is required because https://github.com/caskdata/cdap/blob/develop/cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java#L251 here we use conf to get the compression. But by looking at the comment on the method, seems like this is only used for test case though. Will see if can be removed in the next PR when we pass the `TableDescriptor`

albertshau commented on cdap-hbase-spi/src/main/java/co/cask/cdap/spi/hbase/HBaseDDLExecutor.java:
```diff
@@ -17,13 +17,29 @@
+   * @param conf the configurations with which to initialize
+   * @param <T>
+   */
+  <T> void initialize(T conf);
```
I think it's better to pass in some new Context object, as we can add methods to that Context without breaking this interface. That Context would have the getConfiguration() method. We can do this in a separate PR as well, when we add the ability to access configuration properties in the cconf.

sagarkapare replied:
Agree. Will do it as a separate PR.

albertshau reviewed (commented):
a few minor comments

sagarkapare commented:
@albertshau Please take a look again.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java:
```diff
@@ -69,10 +73,11 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
+    try (HBaseDDLExecutor executor = hBaseDDLExecutorFactory.get();
+         HBaseAdmin admin = new HBaseAdmin(hConf)) {
       if (Strings.isNullOrEmpty(namespaceMeta.getConfig().getHbaseNamespace())) {
         try {
```
should create the executor here instead of in the outside try

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DistributedStorageProviderNamespaceAdmin.java:
```diff
@@ -69,10 +73,11 @@ public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreExcep
     // create HBase namespace
     String hbaseNamespace = tableUtil.getHBaseNamespace(namespaceMeta);
-    try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
+    try (HBaseDDLExecutor executor = hBaseDDLExecutorFactory.get();
```
remove this try and put it just where we use it to call HBaseAdmin.hasNamespace() to avoid creating unnecessarily

albertshau commented on cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/util/hbase/HBase96TableUtil.java:
```diff
@@ -84,14 +85,10 @@ public boolean hasNamespace(HBaseAdmin admin, String namespace) throws IOExcepti
   @Override
-  public void createNamespaceIfNotExists(HBaseAdmin admin, String namespace) throws IOException {
-    Preconditions.checkArgument(admin != null, "HBaseAdmin should not be null");
+  public void createNamespaceIfNotExists(HBaseDDLExecutor ddlExecutor, String namespace) throws IOException {
```
Can we remove this method and just call the ddlExecutor's method? It doesn't do anything at this point.

albertshau commented on cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/util/hbase/HBase96TableUtil.java:
```diff
@@ -92,12 +91,10 @@ public void createNamespaceIfNotExists(HBaseDDLExecutor ddlExecutor, String name
   @Override
-  public void deleteNamespaceIfExists(HBaseAdmin admin, String namespace) throws IOException {
-    Preconditions.checkArgument(admin != null, "HBaseAdmin should not be null");
+  public void deleteNamespaceIfExists(HBaseDDLExecutor ddlExecutor, String namespace) throws IOException {
```
similar comment, can we remove this in favor of just using the ddlExecutor to delete?

albertshau commented on cdap-hbase-compat-1.2-cdh5.7.0/src/main/java/co/cask/cdap/data2/util/hbase/HBase12CDH570TableUtil.java:
```diff
@@ -102,17 +104,19 @@ public void deleteNamespaceIfExists(HBaseDDLExecutor ddlExecutor, String namespa
+    Preconditions.checkArgument(ddlExecutor != null, "HBaseDDLExecutor should not be null");
     Preconditions.checkArgument(tableId != null, "Table Id should not be null.");
-    admin.disableTable(nameConverter.toTableName(tablePrefix, tableId));
+    TableName tableName = nameConverter.toTableName(tablePrefix, tableId);
```
just took a look at HTableNameConverter and I don't understand why its a concrete class and not a utility class with static methods. Can't disableTable and enableTable be the same for all versions of hbase? And if so, could we just remove those methods from HBaseTableUtil in favor of the ddl executor?

albertshau reviewed (commented):
some questions

sagarkapare commented:
@albertshau Addressed comments. Please take a look again. Thanks!

albertshau commented on cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseTableFactory.java:
```diff
@@ -169,9 +172,22 @@ public void disablePayloadTable(String tableName) throws IOException {
+    try {
+      TableName tableName = HTableNameConverter.toTableName(cConf.get(Constants.Dataset.TABLE_PREFIX), tableId);
+      ddlExecutor.enableTableIfDisabled(tableName.getNamespaceAsString(), tableName.getQualifierAsString());
+      LOG.debug("TMS Table {} has been disabled", tableName);
```
disabled -> enabled

albertshau reviewed (approved):
one comment on a log message, other lgtm. fix and merge

sagarkapare commented:
Resolved conflicts. Running build https://builds.cask.co/browse/CDAP-DUT5316-9

sagarkapare commented:
Build is green. Squashing and merging.
